### PR TITLE
Fixed partial dispos

### LINKED ISSUE #291: Unable to sort convictions for cases with missing dispositions
## Description
Unable to sort conviction charges, because one of the dates is a string. Closed cases with partial dispositions are creating charges that do not have a disposition

### Current behavior
The current behavior is that charges for without a disposition are being marked as missing disposition and are being skipped in the expungement process, however this is only working when a case does not have any disposition transactions.

See PR #255 for when this behavior was added.

### Expected behavior
Any charge that does not have a disposition should be skipped regardless if the case has disposition transactions. This is needed for closed cases that have disposition transactions for some, but not all charges.

### Stack trace
```
Traceback (most recent call last):
  File "cli.py", line 117, in <module>
    expunged = expunger.run()
  File "/recordexpungPDX/src/backend/expungeservice/expunger/expunger.py", line 55, in run
    self._set_most_recent_convictions()
  File "/recordexpungPDX/src/backend/expungeservice/expunger/expunger.py", line 94, in _set_most_recent_convictions
    self.convictions.sort(key=lambda charge: charge.disposition.date)
TypeError: '<' not supported between instances of 'str' and 'datetime.date'
```

### Possible fix:
One issue is that there is a mutable default in the base_charge initializer.
https://github.com/codeforpdx/recordexpungPDX/blob/master/src/backend/expungeservice/models/charge_types/base_charge.py#L12
This should be replaced with None and then add a guard for creating a disposition.

### This task is done when:
 - Use case fixture 'CASE_WITH_PARTIAL_DISPOS` to create a system test that exposes bug
 - Create unit test that also exposes bug
 - Any charges that do not have dispositions should be tagged as such and skipped by the time analysis.

### PR DESCRIPTION
**Pull request by aemrich**
Resolved #291 by adding convicted function to BaseCharge and setting a default date for dispositions without a date in the Expunger.

The convicted function in Basecharge explicitly checks the disposition ruling to see if is 'convicted'. The recent_conviction function then checks the convicted function and returns False if not convicted, else returns True if disposition.date was more than 10 years ago.

The Expunger then sets a default date of -1000 years for charges without a disposition date.

### CODE REVIEW

File: src/backend/expungeservice/models/charge_types/base_charge.py [hunk 1]
```diff
@@ -27,11 +27,20 @@ def acquitted(self):
         if not self.disposition.ruling:
             return False
         else:
-            return self.disposition.ruling[0:9] != 'Convicted'
+            return self.disposition.ruling[0:9].lower() != 'convicted'
+
+    def convicted(self):
```
[the remark this answers is not in the document]
Comment (wittejm) [reply src/backend/expungeservice/models/charge_types/base_charge.py]: The problem with this implementation is that a charge can have states of `acquitted`, `convicted`, or neither if there is no disposition, and if that's true then the logic you're proposing is invalid. The functions aren't fully redundant.

File: src/backend/expungeservice/expunger/expunger.py [hunk 2]
```diff
@@ -92,7 +94,7 @@ def _set_most_recent_convictions(self):
             self.second_most_recent_conviction = self.convictions[-2]
 
     def _assign_most_recent_charge(self):
-        self.charges.sort(key=lambda charge: charge.disposition.date, reverse=True)
+        self.charges.sort(key=lambda charge: charge.disposition.date if charge.disposition.date else (date_class.today() + relativedelta(years=-1000)), reverse=True)
```
[the remark this answers is not in the document]
Comment (wittejm) [reply src/backend/expungeservice/expunger/expunger.py]: Question re Expunger intended behavior: Does a charge belong in this list if there is no disposition? If there is no disposition then the charge has no date information and can't be sorted by this. It sounlds like you're suggesting that a charge without a ruling doesn't belong in this list. Is that correct?

The change here is meant to cover the edge case in which at least one charge has a disposition and at least one does not. This is exclusive with the edge case that this list has only one charge.  If a Record has only one Charge and it has no Disposition, is it valid to call it the most recent charge? If yes, then this implementation is valid.

[the remark this answers is not in the document]
Comment (NickSchimek) [reply src/backend/expungeservice/expunger/expunger.py]: @wittejm @aemrich Yes, these methods are doing to much and are checking for null.... Thanks for pointing that out. We can still DRY them out a bit by doing this:
```
    def acquitted(self):
        return self.disposition.ruling and self.disposition.ruling[0:9].lower() != 'convicted'

    def convicted(self):
        return self.disposition.ruling and not self.acquitted()
```
However both the methods are still checking for null and that could be resolved by implementing the null object pattern here. 

I'm fine with the current code in its current state, however upon merging I'd like to create a refactoring ticket to dry the code out and remove the check for null.

[the remark this answers is not in the document]
Comment (aemrich) [reply src/backend/expungeservice/expunger/expunger.py]: Ok great, thanks. @NickSchimek I've made your changes to the acquitted and convicted methods -- I'll go ahead and merge this in now.

File: src/backend/expungeservice/models/charge_types/base_charge.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (NickSchimek) [reply src/backend/expungeservice/models/charge_types/base_charge.py]: @wittejm @aemrich whooops originally put this comment in the wrong spot. Moving it here.
Yes, these methods are doing to much and are checking for null.... Thanks for pointing that out. We can still DRY them out a bit by doing this: 
```
    def acquitted(self):
        return self.disposition.ruling and self.disposition.ruling[0:9].lower() != 'convicted'

    def convicted(self):
        return self.disposition.ruling and not self.acquitted()
```
However both the methods are still checking for null and that could be resolved by implementing the null object pattern here. 

I'm fine with the current code in its current state, however upon merging I'd like to create a refactoring ticket to dry the code out and remove the check for null.

File: src/backend/expungeservice/expunger/expunger.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (NickSchimek) [reply src/backend/expungeservice/expunger/expunger.py]: Thanks for your work on this @aemrich 
In regards to @wittejm questions:
> Does a charge belong in this list if there is no disposition?

No a charge without a disposition should not be in this list. The most recent charge is used in the Time Analyzer to determine the time eligibility for some class b felonies. If a charge without a disposition is put into this list and falls through to the time analyzer then the error is repeated there.

> If there is no disposition then the charge has no date information and can't be sorted by this. It sounds like you're suggesting that a charge without a ruling doesn't belong in this list. Is that correct?

Yes, a charge without a disposition, hence no ruling/date, should not be in this list; because it cannot be evaluated correctly and should be flagged so that we can bring it to the user's attention.

> The change here is meant to cover the edge case in which at least one charge has a disposition and at least one does not.

Not necessarily. That was the case issue provided to expose the bug, hence the reason I also requested unit tests to be added in the original issue. And the case detail `CASE_WITHOUT_DISPOS` can be used to expose the bug too. 

> If a Record has only one Charge and it has no Disposition, is it valid to call it the most recent charge?

No, it should be labeled as the most recent charge, because we could not run time analysis using a charge without a disposition date. Therefore we need to flag the charge and mark it as not having a disposition (as mentioned in the issue) so that the user can be notified to evaluate independently.
